synapse/handlers/oidc_handler.py

```diff
@@ -214,18 +215,27 @@ def _validate_metadata(self):
                 else:
                     raise ValueError('"jwks_uri" must be set')
 
-    @property
-    def _uses_userinfo(self) -> bool:
+    def _uses_userinfo(self, token) -> bool:
         """Returns True if the ``userinfo_endpoint`` should be used.
 
-        This is based on the requested scopes: if the scopes include
-        ``openid``, the provider should give use an ID token containing the
-        user informations. If not, we should fetch them using the
-        ``access_token`` with the ``userinfo_endpoint``.
+        This is based on user_profile_method: if it is ``id_token``,
+        the provider should use an ID token containing the
+        user informations. If it is ``userinfo_endpoint``, we should
+        fetch them using the ``access_token`` with the ``userinfo_endpoint``.
+        If it is ``auto``, we use the ``id_token`` if present in the token
+        response, and fallback to the ``userinfo_endpoint``.
```

Comment:
You don't really need to describe every step of a function in words: it's easier for me to read the code if that's what I want.

Please try to summarise the behaviour of the function instead, and *document the arguments*.

However, I suspect that we don't really need this method at all, and it would make more sense to inline its logic in the one place you need it.

Reply:
This is a doc string not a comment so I tried to match the detail of the old code.
I agree that this is too detailed.

I think we should keep the method though. The old implementation was hardly longer than the method call but having it in a method still felt right. That way, we have a single point of responsibility for deciding whether to use the userinfo endpoint.

Unfortunately, we don't have this with the current code because part of the logic bleeds into `_validate_metadata`. This is another one of the ugly parts of this PR.